### DISCUSSION TITLE
Alpine and Linux Platform Packages

### DISCUSSION THREAD

**Discussion by lukik** [dashingsoft/pyarmor#1870] [Q&A]
The scripts that I'm obfuscating are to run in three platforms. Windows, Linux and Alpine. I've generated the obfuscated script using 
a python script and I've passed the necessary platforms that I need

`['gen', '--platform', 'alpine.x86_64,linux.x86_64,windows.x86_64', '-O', 'dist/obfuscated_files', foo.py']`

This generates several files in the dist directory:

```
└── obfuscated_files
    ├── foo.py
    └── pyarmor_runtime_000000
        ├── __init__.py
        ├── alpine_x86_64
        │   └── pyarmor_runtime.so
        ├── linux_x86_64
        │   └── pyarmor_runtime.so
        └── windows_x86_64
            └── pyarmor_runtime.pyd
```

I am using a fully fledged Aline Server and Not a docker container.

To run the files, I copy over the application source code `foo.py`, the `pyarmor_runtime_000000/__init__.py` files and:

**Attempt 1:**
I copy across the alpine package to the alpine server:

`dist/obfuscated_files/pyarmor_runtime_000000/alpine_x86_64/pyarmor_runtime.so`

- The script fails with:  `ModuleNotFoundError: No module named 'dist.pyarmor_runtime_000000.linux_x86_64'`

**Attempt 2:**
I remove the alpine package in attempt 1 above and I copy over the linux plaform one i.e. `dist/obfuscated_files/pyarmor_runtime_000000/linux_x86_64/pyarmor_runtime.so`

- The script fails with: `ImportError: Error relocating dist/pyarmor_runtime_000000/linux_x86_64/pyarmor_runtime.so: __snprintf_chk: symbol not found`

**Attempt 3:** 
I remove the linux package in attempt 2 above and copy across the alpine package (the one in attempt 1) but then I rename it to linux i.e.

`mv dist/obfuscated_files/pyarmor_runtime_000000/alpine_x86_64  dist/obfuscated_files/pyarmor_runtime_000000/linux_x86_64`

- The script runs!

So, two questions come up here:
1. Does pyarmor think my alpine server is a linux box? if yes, what is it that it looks for to make that call? 
2. If it really thinks it's a linux box, why would it work with the underlying alpine compilation? I would expect it to fail? Or is the alpine compilation "compatible" with linux and so it just runs?

I did review the `pyarmor_runtime_000000/__init__.py` script and it looks like it's the one determining the platform to look for?

**Comment by jondy** [comment]
There are 3 lines in the `pyarmor_runtime_000000/__init__.py` to check this platform is alpine or not:
```python
cname, cver = platform.libc_ver()
                if cname == 'musl':
                    plat = 'alpine'
```
